File: joss/comparisons/cat-6-contextual-runtime/29-empty-dataframe-ops/example-empty-dataframe-ops.py
"""
Error Class 29: Empty DataFrame Operations

Python silently allows operations on empty DataFrames,
producing empty results that may mask bugs.
"""
import pandas as pd

# Empty DataFrame
empty = pd.DataFrame()

# SILENT: Operations on empty DataFrame produce empty results — no warning
result = empty.groupby("x").sum() if "x" in empty.columns else pd.DataFrame()
print(f"empty groupby result: {len(result)} rows")  # 0 rows, no error

# SILENT: Aggregation on empty Series
empty_with_cols = pd.DataFrame({"x": pd.Series(dtype=float)})
total = empty_with_cols["x"].sum()
print(f"sum of empty: {total}")  # 0.0 — not NaN, not error

File: joss/comparisons/cat-6-contextual-runtime/29-empty-dataframe-ops/example-empty-dataframe-ops.R
# Error Class 29: Empty DataFrame Operations
#
# R silently allows operations on empty tibbles,
# producing empty results that may mask bugs.

library(tidyverse)

# Empty tibble
empty <- tibble()

# SILENT: Operations on empty tibble produce empty results
# group_by on empty tibble — no error
empty_with_cols <- tibble(x = numeric())
result <- empty_with_cols %>% group_by(x) %>% summarise(total = sum(x))
cat("empty summarise result:", nrow(result), "rows\n")  # 0 rows, no error

# SILENT: sum of empty numeric — returns 0, not NA, not error
total <- sum(empty_with_cols$x)
cat("sum of empty:", total, "\n")  # 0 — may be unexpected
